src/util/std_types.h

```diff
@@ -1671,4 +1671,66 @@ inline complex_typet &to_complex_type(typet &type)
   return static_cast<complex_typet &>(type);
 }
 
+class function_typet:public typet
+{
+public:
+  function_typet():typet(ID_function)
+  {
+    subtypes().resize(2);
+  }
+
+  class variablet:public irept
+  {
+  public:
+    irep_idt get_identifier() const
+    {
+      return get(ID_identifier);
+    }
+
+    void set_identifier(const irep_idt &identifier)
+    {
+      return set(ID_identifier, identifier);
+    }
+
+    typet &type()
+    {
+      return (typet &)add(ID_type);
+    }
+
+    const typet &type() const
+    {
+      return (const typet &)find(ID_type);
+    }
+  };
+
+  using variablest=std::vector<variablet>;
+
+  variablest &variables()
+  {
+    return (variablest &)subtypes()[0].subtypes();
+  }
+
+  const variablest &variables() const
+  {
+    return (const variablest &)subtypes()[0].subtypes();
+  }
+
+  variablet &add_variable()
+  {
+    auto &v=variables();
+    v.push_back(variablet());
+    return v.back();
+  }
+
+  typet &range()
```

Comment:
What is this meant to describe?

Reply:
The "return type" of the function. It would be technically more correct to say "codomain", it's just that nobody knows that word.

Reply:
I'll switch to codomain; the users of this type should know (or learn) that word.